File: Problems/Algorithms/117. Populating Next Right Pointers in Each Node II/populate_next_pointers_II.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root: return None

        node = root
        while node:
            self.link_children(node)
            node = node.next

        self.connect(root.left)
        self.connect(root.right)

        return root

    def find_next(self, node: 'Node') -> 'Node':
        if not node: return None

        next = node.next
        while next:
            if next.left: return next.left
            if next.right: return next.right
            next = next.next

        return None

    def link_children(self, node: 'Node'):
        if not node: return

        if node.left:
            node.left.next = node.right if node.right else self.find_next(node)

        if node.right:
            node.right.next = self.find_next(node)

File: Problems/Algorithms/117. Populating Next Right Pointers in Each Node II/populate_next_pointers_II.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */

func findNext(node *Node) *Node {
    if node == nil {
        return nil
    }

    next := node.Next
    for next != nil {
        if next.Left != nil {
            return next.Left
        }
        if next.Right != nil {
            return next.Right
        }
        next = next.Next
    }

    return nil
}

func linkChildren(node *Node) {
    if node == nil {
        return
    }

    if node.Left != nil {
        if node.Right != nil {
            node.Left.Next = node.Right
        } else {
            node.Left.Next = findNext(node)
        }
    }

    if node.Right != nil {
        node.Right.Next = findNext(node)
    }
}

func connect(root *Node) *Node {
    if root == nil {
        return nil
    }

    node := root
    for node != nil {
        linkChildren(node)
        node = node.Next
    }

    connect(root.Left)
    connect(root.Right)

    return root
}
